### ISSUE TITLE
Trying to use Authorized Endpoints does create cert correctly?

### ISSUE THREAD

**Issue by bearmoo-repo** [rancher/rancher#18901]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
Have an existing cluster
edit cluster
Enable Authorized Cluster Endpoints
Fillin FQDN
Add cert
click on save

**Result:**
apiVersion: v1
kind: Config
clusters:
- name: "test-cluster"
  cluster:
    server: "https://192.168.0.8/k8s/clusters/c-aa"
    api-version: v1
    certificate-authority-data: 
     "afasdfasdfasdft"
     "..."
     "jljljlj="
- name: "test-cluster-node-02"
  cluster:
    server: "https://192.168.0.13/k8s/clusters/c-aa"
    api-version: v1
    certificate-authority-data: "-----BEGIN CERTIFICATE----- \"
    "ioojlnlknlkjojoij-----END CERTIFICATE-----"
    "...."
users:
- name: "user-xx"
  user:
    token: "kubeconfig-user-xx:yyy"

contexts:
- name: "test-cluster"
  context:
    user: "user-xx"
    cluster: "test-cluster"

current-context: "test-cluster"

**Other details that may be helpful:**

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
- Installation option (single install/HA):
2.2.0-rc4
Single install

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):
Custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory):
cloud
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", Bu
ildDate:"2019-02-01T20:08:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", Bu
ildDate:"2018-12-13T10:31:33Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
v1.13.1
```

**Comment by erikwilson** [comment]
Hi @bearmoo-repo, it looks like you did not fill out the FQDN, because I do not see an FQDN context generated. Could you please show the advanced options setting for the cluster? eg:

<img width="1624" alt="Screen Shot 2019-03-14 at 4 28 28 PM" src="https://user-images.githubusercontent.com/1999874/54398171-40472380-4676-11e9-9536-4e170a3485cc.png">

Thanks!

**Comment by cjellick** [comment]
@erikwilson he also said that this was a "custom" cluster. Those are technically RKE clusters, but they're a little different from rke clusters provisioned via a node driver. Would that have any impact on this feature working?

**Comment by erikwilson** [comment]
@cjellick at the point where we decide to provide fqdn or node contexts for kubeconfig we don't care about the type of cluster (https://github.com/rancher/rancher/blob/master/pkg/kubeconfig/kubeconfig.go#L120). 

We also try to validate the cluster type but are not very restrictive so that tests would work (https://github.com/rancher/rancher/blob/master/pkg/api/customization/cluster/validator.go#L39).

However it is unlikely that a custom cluster would have kube-api-auth running, so the next step of failure might be trying to use one of those contexts. I think we should maybe provide the Authorized Cluster Endpoint UI for RKE clusters only, and if a person does have it working on a custom cluster they can always add a context to the kubeconfig manually.

**Comment by superseb** [comment]
Besides this, the main issue here is that the provided CA certificate is not base64 encoded in the generated kubeconfig.

**Comment by erikwilson** [comment]
That is a good point, thank you @superseb. The original intention was that the UI would send the data to the API already encoded, and possibly being able to handle encoded data pasted into it from another kubeconfig. I think there was some information lost there where we need to revisit the design and testing.

**Comment by cjellick** [comment]
So does this feature not work at all with certificates as its currently impletemend? @erikwilson @bmdepesa @superseb

**Comment by cjellick** [comment]
@bearmoo-repo thaanks for opening this issue. We are wokring on a fix and it will be in the release

**Comment by bearmoo-repo** [comment]
Thanks!

**Comment by erikwilson** [comment]
Yes, thank you @bearmoo-repo! And to clarify my misunderstanding, Authorized Endpoints should work well with Custom clusters, they are integrated with RKE and Rancher.

**Comment by cjellick** [comment]
Fixed this in two ways:
1. on the backend, when we create teh kubeconfig, we will base64 encode the cert, as we should be doing
2. we added validation that if you set the cacert, you must provide an fqdn

**Comment by bmdepesa** [comment]
rancher/rancher:master 3/18

* Created Rancher server
* Created custom cluster

Local cluster auth with no FQDN / no CA
* Enabled auth
* Kubeconfig generated with both contexts and functional
* Embedded kubectl functional
* CA is base64 encoded in API

Local cluster auth with CA / no FQDN
* Errors correctly

Local cluster auth with FQDN & CA
* Created LB with TLS termination at LB at FQDN
* Added custom cluster controlplane node to LB
* Enabled auth with FQDN & CA
* CA is base64 encoded
* Kubeconfig generated with both contexts and functional
* Embedded kubectl functional

Cluster auth with FQDN / no CA
* Enabled auth with FQDN & no CA
* Kubeconfig generated with both contexts and functional
* Embedded kubectl functional

**Comment by bearmoo-repo** [comment]
It looked like the code got merged. I tryied it out, and I still see this:
The version I'm using is 2.2.0-rc6

.kube/config:
-----
apiVersion: v1
kind: Config
clusters:
- name: "kubernetes-example-net-fqdn"
  cluster:
    server: "https://192.168.0.10/k8s/clusters/c-pbh5w"
    certificate-authority-data: "ssuuuuuuukljaldfjaldjfl;adjfal;"
- name: "kubernetes-example-net-fqdn"
  cluster:
    server: "https://kubernetes.example.net"
    certificate-authority-data: "-----BEGIN CERTIFICATE-----
aadfadfadfasdfasddf=\
      -----END\
      CERTIFICATE-----"

users:
- name: "user-xxx"
  user:
    token: "kubeconfig-user-xx-yyyy"

contexts:
- name: "kubernetes-example-net"
  context:
    user: "user-xxx"
    cluster: "kubernetes-example-net"
- name: "kubernetes-example-net-fqdn"
  context:
    user: "user-xxx"
    cluster: "kubernetes-example-net-fqdn"

current-context: "kubernetes-example-net"

**Comment by erikwilson** [comment]
@bearmoo-repo Those changes should now be available in 2.2.0-rc7 if you would like to try it out
